src/extra/gpkg/read.jl

```diff
@@ -0,0 +1,204 @@
+# SHALL be the name of a column in the table or view specified by the table_name
+# column value for that row.
+#
+# Requirement 25: The geometry_type_name value in a gpkg_geometry_columns row
+# SHALL be one of the uppercase geometry type names specified
+
+# Requirement 26: The srs_id value in a gpkg_geometry_columns table row
+# SHALL be an srs_id column value from the gpkg_spatial_ref_sys table.
+#
+# Requirement 27: The z value in a gpkg_geometry_columns table row SHALL be one
+# of 0, 1, or 2.
+#
+# Requirement 28: The m value in a gpkg_geometry_columns table row SHALL be one
+# of 0, 1, or 2.
+#
+# Requirement 146: The srs_id value in a gpkg_geometry_columns table row
+# SHALL match the srs_id column value from the corresponding row in the
+# gpkg_contents table.
+function gpkggeoms(db, ; layer=1)
+  tb = DBInterface.execute(
+    db,
+    """
+SELECT g.table_name AS tn, g.column_name AS cn, c.srs_id as crs, g.z as elev, srs.organization as org, srs.organization_coordsys_id as org_coordsys_id,
+( SELECT type FROM sqlite_master WHERE lower(name) = lower(c.table_name) AND type IN ('table', 'view')) AS object_type
+FROM gpkg_geometry_columns g, gpkg_spatial_ref_sys srs
+JOIN gpkg_contents c ON ( g.table_name = c.table_name )
+WHERE c.data_type = 'features'
+AND (SELECT type FROM sqlite_master WHERE lower(name) = lower(c.table_name) AND type IN ('table', 'view')) IS NOT NULL
+AND g.srs_id = srs.srs_id
+AND g.srs_id = c.srs_id
+AND g.z IN (0, 1, 2)
+AND g.m IN (0, 1, 2)
+ LIMIT $layer;
+    """
+  )
+  featuretablegeoms = map((row.tn, row.cn, row.org, row.org_coordsys_id) for row in tb) do (tn, cn, org, orgcoordsysid)
+    # get feature geometry from geometry column in feature table
+    gpkgbinary = DBInterface.execute(db, "SELECT $cn FROM $tn;")
+    headerlen = 0
+
+    gpkgblobs = filter(map(NamedTuple, gpkgbinary)) do row
+      !ismissing(getfield(row, Symbol(cn))) # ignore all rows with missing geometries
+    end
+
+    geomcollection = map(gpkgblobs) do blob
+      if blob[1][1:2] != UInt8[0x47, 0x50]
+        @warn "Missing magic 'GP' string in GPkgBinaryGeometry"
+      end
+      io = IOBuffer(blob[1])
+      seek(io, 3)
+      flag = read(io, UInt8)
+      # Note that Julia does not convert the endianness for you.
+      # Use ntoh or ltoh for this purpose.
+      bswap = isone(flag & 0x01) ? ltoh : ntoh
+      srsid = bswap(read(io, Int32))
+      envelope = (flag & (0x07 << 1)) >> 1
+      envelopedims = 0
+      if !iszero(envelope)
+        if isone(envelope)
+          envelopedims = 1 # 2D
+        elseif isequal(2, envelope)
+          envelopedims = 2 # 2D+Z
+        elseif isequal(3, envelope)
+          envelopedims = 3 # 2D+M is not supported
+        elseif isequal(4, envelope)
+          envelopedims = 4 # 2D+ZM is not supported
+        else
+          throw(ErrorException("exceeded dimensional limit for geometry, file may be corrupted or reader is broken"))
+        end
```

Comment:
Can you please comment on why the envelopedims is 1 dimension smaller than the dimension of geometries? Your code implies that envelopedims = 2 is equivalent to 2D + z, which I am interpreting as (x, y, z) coordinates.